### DISCUSSION TITLE
Adding linting for consistent coding standards

### DISCUSSION THREAD

**Discussion by adityakiran1423** [NCAR/music-box#290] [Ideas]
Hello everyone, this discussion is regarding linting, more specifically adding linting to the CI process.
There are a lot of projects going on in C++/C# which has led to contributors involved in both projects using the same conventions in the python codebase. 
For example, check issue [#286](https://github.com/NCAR/music-box/issues/286), adding linting will ensure this issue won't occur again.

If we go ahead with adding linting support into the CI process, there are two main options to choose from :
1. [pylint](https://pypi.org/project/pylint/)
2. [flake8](https://pypi.org/project/flake8/)

Both of them have their cons and pros, this table provides a good reference ([link to adrticle](https://dsstream.com/improve-your-python-code-quality/)): 

![image](https://github.com/user-attachments/assets/1199efbc-f852-4ca1-bf2c-a7d97cacd8e0)

Any ideas, questions, suggestions or queries regarding this can be sorted and decided whether linting can be added or not.

Thank you!!

**Comment by K20shores** [comment]
@boulderdaze @montythind @mattldawson @mwaxmonsky do any of you have thoughts on this?

**Reply by mwaxmonsky** [reply]
Fully support adding these in!

Another one I've seen referenced a fair bit recently that I would like to experiment with that's really taking off is [`ruff`](https://github.com/astral-sh/ruff) that looks to have some of these built in.

The different linters do have some conflicting rules that we'd have to figure out but no harm in adding multiple linters and dealing with the conflicting overlap when needed (that may change if one get's too conflicting with another but the last time I ran these concurrently, it was more the exception than the norm).

Some of the formatting options (missing newline, missing whitespaces, etc.) can be covered by [yapf](https://github.com/google/yapf) (which I don't recommend from past experiences) and Python Black (which I've used before and have had no real issues with) but then we cede configuration of the format to it's standard (ie, we might not be able to change how it formats something if we don't like it.).

Also I think flake8 incorporates pyflakes so we will get it's checks for free with pyflakes.

**Reply by K20shores** [reply]
We already have an action that formats using [autopep8](https://github.com/NCAR/music-box/blob/main/.github/workflows/pep8_autoformat.yml). How would the tools mentioned by @adityakiran1423 and what you've mentioned(@mwaxmonsky) add more value?

**Reply by mwaxmonsky** [reply]
The `autopep8` looks to be a partial formatter (just the pep8 guidelines), but Black and yapf are  total source code formatters (things that the standards don't have rules for).  This would be similar to clang-format for our C++ code.

The linters (pylint, flake8, etc) add code-smell detection like clang-tidy.

Ruff looks to blends all of them together under one tool that handles the conflicting settings for us.

The formatters might not add much value (especially if we have different preferences from the built in config) but I think the linters would be very valuable to add to the workflow to catch things we might miss.

**Reply by K20shores** [reply]
Got it. I like the sound of ruff simply because it mixes the linting and formatting. @adityakiran1423 would you be interested in rewriting our pep8 action so that it uses ruff? If so, I'll make an issue with a few suggestions

**Reply by adityakiran1423** [reply]
Hello, apologies for the delayed message. I wouldn't mind taking up rewriting the pep8 action @K20shores !! I'll have to go through their docs first. As to what @mwaxmonsky mentioned about conflicting rules, I think ruff has a config file in which we can specify the the style guide we want to implement. We can make ruff use the black formatter from what I have heard. Just wondering, for what other action do we need to use anything apart from ruff @mwaxmonsky ?
Here's the page which explains [configuring ruff](https://docs.astral.sh/ruff/configuration/)

**Reply by K20shores** [reply]
We don't have any particular rules that we're looking to enforce from the ACOM side. @mwaxmonsky, is there anything from CAM-SIMA (which is sensible) that we should enforce in our python code?

**Reply by mwaxmonsky** [reply]
Apologies for the delay!  We currently have a `.pylintrc` in CAM-SIMA (https://github.com/ESCOMP/CAM-SIMA/blob/development/test/.pylintrc) but as this is only the `pylint` options (`ruff` will have many more options from the different analyzers), I'm amenable to any configuration options that are reasonable.

**Reply by mwaxmonsky** [reply]
I also spent some time digging into our other core projects and we had a linter configured in https://github.com/NCAR/ccpp-framework but I had to remove it because it was causing conflicts (particularly between python 3.8 and 3.9+).  It's been on my todo list to add it back to that as well.
